support UNS in hadoop daos

jiafuzha opened the pull request:
Signed-off-by: jiafu zhang <jiafu.zhang@intel.com>

jiafuzha committed:
support UNS in hadoop daos

Signed-off-by: jiafu zhang <jiafu.zhang@intel.com>

minmingzhu commented on src/client/java/hadoop-daos/src/main/java/io/daos/fs/hadoop/DaosFileSystem.java:
```diff
@@ -158,10 +178,120 @@ public void initialize(URI name, Configuration conf)
+    if (!"POSIX".equalsIgnoreCase(info.getLayout())) {
+      throw new IllegalArgumentException("expect POSIX file system, but " + info.getLayout());
+    }
+    String poolId = info.getPoolId();
+    String contId = info.getContId();
```
@zjf2012  I thought poolid and contid need to check whether null in this position. At the same time , I also verified it in DaosUns.getAccessInfo() that was not  judgment poolid and contid whether null .

jiafuzha committed:
support uns path and container path in on URI

Signed-off-by: jiafu zhang <jiafu.zhang@intel.com>

jiafuzha committed:
fixed merge conflict

Signed-off-by: jiafu zhang <jiafu.zhang@intel.com>

jiafuzha committed:
add apache copyright

Signed-off-by: jiafu zhang <jiafu.zhang@intel.com>

jiafuzha committed:
more tests for uns

Signed-off-by: jiafu zhang <jiafu.zhang@intel.com>

jiafuzha committed:
fixed issue of failed to list root dir from uns path

Signed-off-by: jiafu zhang <jiafu.zhang@intel.com>

minmingzhu commented on src/client/java/hadoop-daos/src/main/java/io/daos/fs/hadoop/DaosFileSystem.java:
```diff
@@ -238,19 +436,46 @@ public void initialize(URI name, Configuration conf)
+        builder.ranks(svc);
+      }
+      this.daos = builder.build();
+      String tmpUri = name.getScheme() + "://" + name.getAuthority();
+      workPath = "/user/" + System.getProperty("user.name");
+      this.uri = URI.create(tmpUri);
+      if (uns) {
+        qualifiedUnsPrefix = tmpUri + unsPrefix;
+        qualifiedUnsWorkPath = qualifiedUnsPrefix + workPath;
+        workingDir = new Path(qualifiedUnsWorkPath);
+      } else {
+        this.workingDir = new Path(workPath)
+          .makeQualified(this.uri, this.getWorkingDirectory());
+      }
```
@zjf2012  i thought this code should be changed as follows, because, the uri of  daos ubs  should not be http://uns:1  instend of  http://uns:1/unsPrefix
`
      if (uns) {
        qualifiedUnsPrefix = tmpUri + unsPrefix;
        qualifiedUnsWorkPath = qualifiedUnsPrefix + workPath;
        this.uri = URI.create(qualifiedUnsPrefix );
        workingDir = new Path(qualifiedUnsWorkPath);
      } else {
        this.uri = URI.create(tmpUri);
        this.workingDir = new Path(workPath)
          .makeQualified(this.uri, this.getWorkingDirectory());
      }
`

minmingzhu commented on src/client/java/hadoop-daos/src/main/java/io/daos/fs/hadoop/DaosFileSystem.java:
```diff
@@ -158,10 +184,171 @@ public void initialize(URI name, Configuration conf)
-    if (StringUtils.isEmpty(name.getAuthority())) {
-      throw new IllegalArgumentException("authority (ip:port) cannot be empty. we need it to identify a " +
-              "unique DAOS File System.");
+    String authority = name.getAuthority();
+    if (Constants.DAOS_AUTHORITY_UNS.equals(authority)) {
+      throw new IllegalArgumentException("need uns id in authority, like daos://uns:1/...");
+    }
+    if (StringUtils.isEmpty(authority) || authority.startsWith(Constants.DAOS_AUTHORITY_UNS+":")) {
+      LOG.info("initializing from uns path, " + name.getPath());
+      String fields[] = authority.split(":");
+      if (fields.length != 2) {
+        throw new IllegalArgumentException("invalid authority, " + authority);
+      }
+      try {
+        int p = Integer.valueOf(fields[1]);
+        if (p < 1) {
+          throw new IllegalArgumentException("uns id should be no less than 1. " + authority);
+        }
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException("bad uns id. should be integer. " + authority);
+      }
+      uns = true;
+      initializeFromUns(name, conf);
+    } else {
+      LOG.info("initializing from config file");
+      uns = false;
+      initializeFromConfigFile(name, conf);
+    }
```
@zjf2012  We have two types of name resolution, one is daos://default:1 and the other is daos://uns:1. Other forms should throw exceptions, so I think it should be written like this :
`
if (StringUtils.isEmpty(authority) || authority.startsWith(Constants.DAOS_AUTHORITY_UNS+":")) {
...
} else if (authority.startsWith(Constants.DAOS_CONFIG_POOL_KEY_DEFAULT+":")) {
...
} else {
throw new IllegalArgumentException("need uns id in authority, like daos://uns:1/... or daos://default:1");
}
`

minmingzhu committed:
resolve the problem  that path has not scheme and authority (#2952)

* 1. modify uri and workingDir on uns model
2. modify resolvePath method , when path.startsWith(unsPrefix), path add scheme and authority

Signed-off-by: minmingzhu <minming.zhu@intel.com>

* 1. rollback validateAndConnect() function

Signed-off-by: minmingzhu <minming.zhu@intel.com>
